presto-hive/src/main/java/com/facebook/presto/hive/S3SelectPushdown.java

```diff
@@ -0,0 +1,193 @@
+
+    private final HiveClientConfig hiveClientConfig;
+
+    S3SelectPushdown(HiveClientConfig hiveClientConfig)
+    {
+        this.hiveClientConfig = requireNonNull(hiveClientConfig);
+    }
+
+    private static boolean isSerdeSupported(Properties schema)
+    {
+        String serdeName = getDeserializerClassName(schema);
+        return SUPPORTED_SERDES.contains(serdeName);
+    }
+
+    private static boolean isInputFormatSupported(Properties schema)
+    {
+        String inputFormat = getInputFormatName(schema);
+        return SUPPORTED_INPUT_FORMATS.contains(inputFormat);
+    }
+
+    public static boolean isCompressionCodecSupported(InputFormat inputFormat, Path path)
+    {
+        if (inputFormat instanceof TextInputFormat) {
+            try {
+                Optional<CompressionCodec> compressionCodec = getCompressionCodec((TextInputFormat) inputFormat, path);
+                if (!compressionCodec.isPresent() ||
+                        stream(S3SelectSupportedCompressionCodecs.values())
+                                .anyMatch(codec -> codec.getCodecClass().isInstance(compressionCodec.get()))) {
+                    return true;
+                }
+            }
+            catch (IllegalAccessException | NoSuchFieldException e) {
+                LOG.error(e, "Failed to find compressionCodec for inputFormat: %s", inputFormat.getClass().getName());
+                return false;
+            }
+        }
+        return false;
+    }
+
+    private static boolean areColumnTypesSupported(List<Column> columns)
+    {
+        if (columns == null || columns.isEmpty()) {
+            return false;
+        }
+
+        for (Column column : columns) {
+            String type = column.getType().getHiveTypeName().toString();
+            if (type.startsWith(StandardTypes.DECIMAL)) {
+                // skip precision and scale when check decimal type
+                type = StandardTypes.DECIMAL;
+            }
+            if (!SUPPORTED_COLUMN_TYPES.contains(type.toLowerCase(Locale.ENGLISH))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean isS3SelectEnabled(ConnectorSession session)
+    {
+        if (!hiveClientConfig.isS3SelectPushdownEnabled()) {
+            return false;
+        }
+        return HiveSessionProperties.isS3SelectPushdownEnabled(session);
+    }
+
+    private static boolean isS3Storage(Properties schema)
+    {
+        String location = schema.getProperty(META_TABLE_LOCATION);
+        return location != null && SUPPORTED_S3_PREFIXES.stream().anyMatch(location::startsWith);
+    }
+
+    boolean shouldOptimizeTable(ConnectorSession session, Table table, Optional<Partition> partition)
+    {
+        if (!isS3SelectEnabled(session)) {
+            return false;
+        }
+        // Hive table partitions could be on different storages,
+        // as a result, we have to check each individual partition
+        Properties schema;
```

Comment:
This could be written as
```java
Properties schema = optionalPartition
        .map(partition -> getHiveSchema(partition, table))
        .orElseGet(() -> getHiveSchema(table);